src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_visitor.h

```diff
@@ -0,0 +1,233 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "node_extent_manager.h"
+#include "node_delta_recorder.h"
+#include "node_layout_replayable.h"
+
+namespace crimson::os::seastore::onode {
+
+template <typename FieldType, node_type_t NODE_TYPE>
+class DeltaRecorderT final: public DeltaRecorder {
+ public:
+  using layout_t = NodeLayoutReplayableT<FieldType, NODE_TYPE>;
+  using position_t = typename layout_t::position_t;
+  using StagedIterator = typename layout_t::StagedIterator;
+  using value_t = typename layout_t::value_t;
+  static constexpr auto FIELD_TYPE = layout_t::FIELD_TYPE;
+
+  ~DeltaRecorderT() override = default;
+
+  template <KeyT KT>
+  void encode_insert(
+      const full_key_t<KT>& key,
+      const value_t& value,
+      const position_t& insert_pos,
+      const match_stage_t& insert_stage,
+      const node_offset_t& insert_size) {
+    // TODO encode to encoded
+  }
+
+  void encode_split(
+      const StagedIterator& split_at,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  template <KeyT KT>
+  void encode_split_insert(
+      const StagedIterator& split_at,
+      const full_key_t<KT>& key,
+      const value_t& value,
+      const position_t& insert_pos,
+      const match_stage_t& insert_stage,
+      const node_offset_t& insert_size,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  void encode_update_child_addr(
+      const laddr_t new_addr,
+      const laddr_packed_t* p_addr,
+      const char* p_start) {
+    // TODO encode to encoded
+  }
+
+  static DeltaRecorderURef create() {
+    return std::unique_ptr<DeltaRecorder>(new DeltaRecorderT());
+  }
+
+ private:
+  DeltaRecorderT() = default;
+  node_type_t node_type() const override { return NODE_TYPE; }
+  field_type_t field_type() const override { return FIELD_TYPE; }
+  void apply_delta(ceph::bufferlist::const_iterator& delta,
+                   NodeExtentMutable& node) override {
+    assert(is_empty());
+    // TODO decode and apply
+    assert(false && "not implemented");
+  }
+};
+
+template <typename FieldType, node_type_t NODE_TYPE>
+class NodeExtentT {
+ public:
+  using layout_t = NodeLayoutReplayableT<FieldType, NODE_TYPE>;
+  using node_stage_t = typename layout_t::node_stage_t;
+  using position_t = typename layout_t::position_t;
+  using recorder_t = DeltaRecorderT<FieldType, NODE_TYPE>;
+  using StagedIterator = typename layout_t::StagedIterator;
+  using value_t = typename layout_t::value_t;
+  static constexpr auto FIELD_TYPE = layout_t::FIELD_TYPE;
+  enum class state_t {
```

Comment:
This should not be recorded independently, can you simply infer it from the state on the underlying CachedExtentRef?

Reply:
`NodeExtentT` is more abstracted than `CachedExtent`. This class only cares about whether or not to record deltas, or need to mutate the underlying `NodeExtentRef` before recording. That's why there're only 3 distinct states in `NodeExtentT`. The transitions between these states are implementation details and are asserted carefully in case of any unexpected behavior changes from the underlying `CachedExtent`.

Reply:
You could wrap the CachedExtentRef check with a helper here.  These states really do appear to map directly onto CachedExtent state.  Is the problem that the dummy backend doesn't provide that kind of tracking?

Reply:
NVM, I'm fine to remove `state_t` and use the underlying states of `CachedExtent` directly with some helpers.

Reply:
fixed in https://github.com/ceph/ceph/pull/36986/commits/2f8e054562bc27d269a90b48aa34f1aeef77e3df